test/gpu/native/studies/shoc/sort.chpl

```diff
@@ -0,0 +1,679 @@
+            __primitive("gpu syncThreads");
+
+        }
+        for j in 0..3:uint(32) {
+            keysOut[4 * i + j] = key[j];
+            valuesOut[4 * i + j] = value[j];
+        }
+    }
+}
+
+proc scan4(idata : 4*uint(32), ref ptr : c_ptr(uint(32)), threadId : uint(32)){
+
+    var val4: 4*uint(32) = idata;
+    var sum: 4*uint(32);
+
+    // Scan the 4 elements in idata within this thread
+    sum[0] = val4[0];
+    sum[1] = val4[1] + sum[0];
+    sum[2] = val4[2] + sum[1];
+    var val:uint(32) = val4[3] + sum[2];
+
+    // Now scan those sums across the local work group
+    val = scanLSB(val, ptr, threadId);
+
+    val4[0] = val;
+    val4[1] = val + sum[0];
+    val4[2] = val + sum[1];
+    val4[3] = val + sum[2];
+
+    return val4;
+}
+
+proc scanLSB(const val:uint(32), ref s_data : c_ptr(uint(32)), threadId: uint(32)){
+    // Shared memory is 256 uints long, set first half to 0's
+    // var idx = __primitive("gpu threadIdx x"): uint(32);
+    var idx = threadId;
+    s_data[idx] = 0 : uint(32);
+    __primitive("gpu syncThreads");
+
+    // Set second half to thread local sum i.e.
+    // the sum of the 4 elems from global mem
+    idx += SORT_BLOCK_SIZE : uint(32); // 128 in this case
+
+    // Unrolled scan in local memory
+    var t : uint(32);
+    s_data[idx] = val;     __primitive("gpu syncThreads");
+    t = s_data[idx -  1];  __primitive("gpu syncThreads");
+    s_data[idx] += t;      __primitive("gpu syncThreads");
+    t = s_data[idx -  2];  __primitive("gpu syncThreads");
+    s_data[idx] += t;      __primitive("gpu syncThreads");
+    t = s_data[idx -  4];  __primitive("gpu syncThreads");
+    s_data[idx] += t;      __primitive("gpu syncThreads");
+    t = s_data[idx -  8];  __primitive("gpu syncThreads");
+    s_data[idx] += t;      __primitive("gpu syncThreads");
+    t = s_data[idx - 16];  __primitive("gpu syncThreads");
+    s_data[idx] += t;      __primitive("gpu syncThreads");
+    t = s_data[idx - 32];  __primitive("gpu syncThreads");
+    s_data[idx] += t;      __primitive("gpu syncThreads");
+    t = s_data[idx - 64];  __primitive("gpu syncThreads");
+    s_data[idx] += t;      __primitive("gpu syncThreads");
+
+    return s_data[idx] - val;  // convert inclusive -> exclusive
+}
+
+
+proc findRadixOffsets(findGlobalWorkSize, ref keys : [] uint(32), ref counters : [] uint(32),
+        ref blockOffsets : [] uint(32), startbit :uint(32), numElements :uint(32), totalBlocks :uint(32) ){
+
+    forall i in 0..<findGlobalWorkSize:uint(32){
+
+        __primitive("gpu set blockSize", SCAN_BLOCK_SIZE);
+
+        var startVoidPtr = __primitive("gpu allocShared", 16*numBytes(uint(32)));
+        var sStartPointers = startVoidPtr : c_ptr(uint(32));
+
+        var sRadixVoidPtr = __primitive("gpu allocShared", 2*SCAN_BLOCK_SIZE*numBytes(uint(32)));
+        var sRadix1 = sRadixVoidPtr : c_ptr(uint(32));
+
+        // var groupId = __primitive("gpu blockIdx x") :uint(32);
```

Comment:
You can take out these commented calls to `__primitive`, I don't think we'll ever support users getting the blockIdx and threadIdx directly from primitives like this.